srml/contract/src/gas.rs

```diff
@@ -236,17 +232,22 @@ pub fn buy_gas<T: Trait>(
 
 /// Refund the unused gas.
 pub fn refund_unused_gas<T: Trait>(transactor: &T::AccountId, gas_meter: GasMeter<T>) {
+	let gas_spent = gas_meter.spent();
+	let gas_left = gas_meter.gas_left();
+
 	// Increase total spent gas.
 	// This cannot overflow, since `gas_spent` is never greater than `block_gas_limit`, which
 	// also has T::Gas type.
-	let gas_spent = <Module<T>>::gas_spent() + gas_meter.spent();
-	<GasSpent<T>>::put(gas_spent);
+	<GasSpent<T>>::mutate(|block_gas_spent| *block_gas_spent += gas_spent);
 
 	// Refund gas left by the price it was bought.
 	let b = <balances::Module<T>>::free_balance(transactor);
-	let refund = <T::Gas as As<T::Balance>>::as_(gas_meter.gas_left) * gas_meter.gas_price;
+	let refund = <T::Gas as As<T::Balance>>::as_(gas_left) * gas_meter.gas_price;
+	let fee = <T::Gas as As<T::Balance>>::as_(gas_spent) * gas_meter.gas_price;
+
+	// this should be infallible since we just charged for it this block.
+	let _ = T::GasPayment::on_unbalanced_decrease(fee);
```

Comment:
Shouldn't the comment be just `expect` proof?

Reply:
nothing should ever panic in the runtime. better to fail and allow a chain to be rescued through governance in subsequent blocks than brick it completely.